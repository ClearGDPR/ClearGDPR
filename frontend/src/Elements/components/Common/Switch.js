import React from 'react';
import PropTypes from 'prop-types';
import styles from '../../theme/Switch.scss';

class Switch extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      value: props.value
    };
  }

  componentDidUpdate() {
    this.setState({ value: this.props.value });
  }

  handleClick() {
    if (this.props.disabled) {
      return;
    }
    this.setState(prevState => ({
      value: !prevState.value
    }));

    if (this.props.onChange) {
      this.props.onChange(!this.state.value);
    }
  }

  render() {
    const classes = [styles.switch, this.state.value ? styles.switchOn : ''].join(' ');

    return (
      <div className={classes} onClick={this.handleClick.bind(this)}>
        <div className={styles.switchToggle} />
      </div>
    );
  }
}

Switch.propTypes = {
  onChange: PropTypes.func.isRequired,
  value: PropTypes.bool,
  disabled: PropTypes.bool
};

export default Switch;
