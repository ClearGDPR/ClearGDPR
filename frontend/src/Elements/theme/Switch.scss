@import 'variables';

.switch {
  border: 1px solid #ccc;
  width: 50px;
  height: 26px;
  border-radius: 13px;
  cursor: pointer;
  display: inline-block;

  &On {
    background: $green;

    .switchToggle {
      left: 23px;
    }
  }

  &Disabled {
    cursor: not-allowed;
  }
}

.switchToggle {
  border: 1px solid #999;
  box-shadow: 1px 1px 1px #ccc;
  width: 25px;
  height: 24px;
  left: 0;
  border-radius: 12px;
  background: white;
  position: relative;
  transition: left 0.2s ease-in-out;
}
