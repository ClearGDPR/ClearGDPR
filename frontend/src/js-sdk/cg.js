import 'es6-promises';
import Config from './config';
import Modules from './modules';
import { SDKError } from './common/exceptions';
import { mergeOptions } from './utils/helpers';

/**
 * Creates a new CG ClearGDPR API client
 * @constructor
 * @classdesc The CG SDK class provides methods to manage user's data
 * @param {Object}  options
 * @param {String}  options.apiKey Key used to sign up to the API
 * @param {String}  options.apiUrl URL to the API server without trailing slash
 * @param {String}  options.apiVersion Version of the API, like `v1`
 * @param {String}  options.apiTimeout Timeout used when a Request is made to the API
 * @param {Boolean} options.debug Flag to enable or disable complete error stack trace
 * @see {@link https://docs.cleargdpr.com/api}
 */
export class CG {
  constructor({ apiKey, apiUrl }) {
    // Enforce accessToken option to be sent.
    if (!apiKey) {
      throw new SDKError('API Key is required.');
    }

    if (!apiUrl) {
      throw new SDKError('API URL is required.');
    }

    this._config = mergeOptions(Config, { apiUrl });
    this.setAPIKey(apiKey);
    this.registerModules();
  }

  registerModules() {
    Object.keys(Modules).map(key => {
      const endpoint = new Modules[key](this);
      Object.assign(this, { [key]: endpoint });
    });
  }

  getAccessToken() {
    return this._config.auth;
  }

  setAccessToken(accessToken) {
    if (accessToken) {
      // TODO: Should validate quicker if its a valid JWT
      // TODO: How to verify JWT with our Server + Consumer
      this._setConfigField('auth', `Bearer ${accessToken}`);
    }
  }

  getAPIKey() {
    return this._config.auth;
  }

  setAPIKey(apiKey) {
    if (apiKey) {
      this._setConfigField('apiKey', apiKey);
    }
  }

  getBaseURL() {
    return this._config.apiVersion
      ? `${this._config.apiUrl}/${this._config.apiVersion}`
      : this._config.apiUrl;
  }

  _setConfigField(key, value) {
    this._config[key] = value;
  }
}
