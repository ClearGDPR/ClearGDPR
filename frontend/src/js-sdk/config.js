const Config = {
  auth: null,
  apiKey: process.env.API_KEY,
  apiUrl: process.env.API_URL,
  apiVersion: process.env.API_VERSION,
  apiTimeout: 30,
  debug: false
};

export default Config;
