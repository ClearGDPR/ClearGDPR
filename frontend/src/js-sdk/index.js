/*
 * cg-js-sdk
 * https://www.cleargdpr.com/
 * 
 */
export { CG } from './cg';
