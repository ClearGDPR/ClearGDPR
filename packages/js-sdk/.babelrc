{
  "presets": [
    [
      "env",
      {
        "targets": {
          "browsers": ["> 1%", "last 2 versions"]
        }
      }
    ]
  ],
  "plugins": [
    [
      "module-resolver",
      {
        "extensions": [".js", ".jsx", ".es", ".es6", ".mjs"],
        "root": ["./src"],
        "alias": {}
      }
    ]
  ]
}
