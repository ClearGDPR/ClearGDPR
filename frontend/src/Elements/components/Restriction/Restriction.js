import React from 'react';
import PropTypes from 'prop-types';
import Switch from '../Common/Switch';
import { map } from 'lodash';

const RESTRICTION_LABELS = {
  directMarketing: 'Direct marketing',
  emailCommunication: 'Email communication',
  research: 'Research and Analytics'
};

class Restriction extends React.PureComponent {
  state = {
    restrictions: {}
  };

  async componentDidMount() {
    const restrictions = await this.props.cg.Subject.getRestrictions();
    this.setState({ restrictions });
  }

  toggleRestriction = async (e, key, value) => {
    const restrictionsPayload = { ...this.state.restrictions, ...{ [key]: !value } };
    this.setState({ restrictions: restrictionsPayload });

    return await this.props.cg.Subject.updateRestrictions(restrictionsPayload);
  };

  render() {
    const { label } = this.props.options;
    const { restrictions } = this.state;

    console.log({ restrictions });

    return (
      <div>
        <label>{label}</label>
        <div
          style={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'left'
          }}
        >
          {map(restrictions, (value, key) => (
            <div
              key={key}
              style={{
                display: 'flex',
                justifyContent: 'space-between'
              }}
            >
              <span>{RESTRICTION_LABELS[key]}</span>
              <Switch onChange={e => this.toggleRestriction(e, key, value)} value={value} />
            </div>
          ))}
        </div>
      </div>
    );
  }
}

Restriction.propTypes = {
  options: PropTypes.object,
  cg: PropTypes.object
};

Restriction.defaultProps = {
  options: {
    label: 'Restrict processing'
  }
};

export default Restriction;
