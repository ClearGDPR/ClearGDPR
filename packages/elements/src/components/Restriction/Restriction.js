import React from 'react';
import PropTypes from 'prop-types';
import Switch from 'components/Common/Switch';
import { map } from 'lodash';
import Subject from '../../contexts/Subject';

const RESTRICTION_LABELS = {
  directMarketing: 'Direct marketing',
  emailCommunication: 'Email communication',
  research: 'Research and Analytics'
};

class Restriction extends React.PureComponent {
  state = {
    restrictions: {}
  };

  componentDidMount() {
    this.props.subject.fetchRestrictions();
  }

  toggleRestriction = (e, key, value) => {
    const restrictionsPayload = { ...this.props.subject.restrictions, ...{ [key]: !value } };
    this.props.subject.updateRestrictions(restrictionsPayload);
  };

  render() {
    const { label } = this.props.options;
    const { restrictions } = this.props.subject;

    return (
      <div>
        <label>{label}</label>
        <div
          style={{
            display: 'flex',
            flexDirection: 'column',
            alignItems: 'left'
          }}
        >
          {map(restrictions, (value, key) => (
            <div
              key={key}
              style={{
                display: 'flex',
                justifyContent: 'space-between'
              }}
            >
              <span>{RESTRICTION_LABELS[key]}</span>
              <Switch onChange={e => this.toggleRestriction(e, key, value)} value={value} />
            </div>
          ))}
        </div>
      </div>
    );
  }
}

Restriction.propTypes = {
  options: PropTypes.object,
  subject: PropTypes.instanceOf(Subject)
};

Restriction.defaultProps = {
  options: {
    label: 'Restrict processing'
  }
};

export default Restriction;
