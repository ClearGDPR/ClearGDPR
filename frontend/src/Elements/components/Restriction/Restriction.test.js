import React from 'react';
import { mount } from 'enzyme';
import Restriction from './Restriction';
import { CG } from '../../../js-sdk';
import Switch from '../Common/Switch';

const restrictionsStub = {
  directMarketing: true,
  emailCommunication: false,
  research: false
};

const cg = new CG({
  apiKey: 'test',
  apiUrl: 'test'
});
cg.Subject.getRestrictions = async () => restrictionsStub;
cg.Subject.updateRestrictions = async () => {};

const setup = () => mount(<Restriction {...{ options: {}, cg }} />);

let spy;

describe('(Elements SDK) Restriction', () => {
  beforeEach(() => {
    spy = null;
  });

  it('should render correctly', () => {
    const component = setup();
    expect(component).toMatchSnapshot();
  });

  it('should call `getRestrictions` method to retrieve current status', async () => {
    spy = jest.spyOn(cg.Subject, 'getRestrictions');
    setup();
    expect(spy).toHaveBeenCalled();
  });

  it('state should be equal to restrictions status', async () => {
    const component = await setup();
    const { restrictions } = component.state();
    expect(restrictions).toEqual(restrictionsStub);
  });

  it('should call `updateRestrictions` when subject toggles switch', async () => {
    spy = jest.spyOn(cg.Subject, 'updateRestrictions');
    const component = await setup();
    component.update();
    component.find(Switch).at(0).simulate('click');
    component.find(Switch).at(1).simulate('click');
    component.find(Switch).at(2).simulate('click');
    expect(component.state().restrictions).toEqual({
      directMarketing: false,
      emailCommunication: true,
      research: true
    });
    expect(spy).toHaveBeenCalledTimes(3);
  });

  afterEach(() => {
    if (spy) {
      spy.mockClear();
    }
  });
});
