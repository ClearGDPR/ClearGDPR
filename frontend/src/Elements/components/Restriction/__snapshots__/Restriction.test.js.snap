// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Elements SDK) Restriction should render correctly 1`] = `
<Restriction
  options={Object {}}
  subject={
    Subject {
      "_propagateMutation": [Function],
      "cg": v {
        "Events": y {
          "_events": Object {},
        },
        "Processors": Processors {
          "_request": s {
            "config": [Circular],
          },
        },
        "Subject": Subject {
          "_request": s {
            "config": [Circular],
          },
          "getRestrictions": [Function],
          "updateRestrictions": [Function],
        },
        "_config": Object {
          "apiKey": "test",
          "apiTimeout": 30,
          "apiUrl": "test",
          "apiVersion": undefined,
          "auth": null,
          "debug": false,
        },
      },
      "data": null,
      "isErased": false,
      "isFetched": false,
      "isGuest": true,
      "isLoading": false,
      "objection": null,
      "processors": null,
      "restrictions": null,
      "shares": null,
      "status": null,
    }
  }
>
  <div>
    <label />
    <div
      style={
        Object {
          "alignItems": "left",
          "display": "flex",
          "flexDirection": "column",
        }
      }
    />
  </div>
</Restriction>
`;
